extensions/pkg/controller/operatingsystemconfig/bash_test.go

```diff
@@ -0,0 +1,172 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operatingsystemconfig_test
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	. "github.com/gardener/gardener/extensions/pkg/controller/operatingsystemconfig"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+)
+
+var _ = Describe("Bash", func() {
+	Describe("#FilesToDiskScript", func() {
+		var (
+			ctx        = context.Background()
+			fakeClient client.Client
+			namespace  = "namespace"
+		)
+
+		BeforeEach(func() {
+			fakeClient = fakeclient.NewClientBuilder().Build()
+		})
+
+		It("should fail when a referenced secret cannot be read", func() {
+			files := []extensionsv1alpha1.File{{
+				Content: extensionsv1alpha1.FileContent{
+					SecretRef: &extensionsv1alpha1.FileContentSecretRef{
+						Name: "foo",
+					},
+				},
+			}}
+
+			script, err := FilesToDiskScript(ctx, fakeClient, namespace, files)
+			Expect(err).To(BeNotFoundError())
+			Expect(script).To(BeEmpty())
+		})
+
+		It("should generate the expected output", func() {
+			Expect(fakeClient.Create(ctx, &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo",
+					Namespace: namespace,
+				},
+				Data: map[string][]byte{"bar": []byte("bar-content")},
+			})).To(Succeed())
+
+			files := []extensionsv1alpha1.File{
+				{
+					Path: "/foo/bar.txt",
+					Content: extensionsv1alpha1.FileContent{
+						SecretRef: &extensionsv1alpha1.FileContentSecretRef{
+							Name:    "foo",
+							DataKey: "bar",
+						},
+					},
+				},
+				{
+					Path: "/bar/baz",
+					Content: extensionsv1alpha1.FileContent{
+						Inline: &extensionsv1alpha1.FileContentInline{
+							Encoding: "",
+							Data:     "plain-text",
+						},
+					},
+				},
+				{
+					Path: "/bar/baz",
+					Content: extensionsv1alpha1.FileContent{
+						Inline: &extensionsv1alpha1.FileContentInline{
+							Encoding: "b64",
+							Data:     "YmFzZTY0",
+						},
+					},
+				},
+				{
+					Path: "/baz/foo",
+					Content: extensionsv1alpha1.FileContent{
+						Inline: &extensionsv1alpha1.FileContentInline{
+							Encoding: "",
+							Data:     "transmit-unencoded",
+						},
+						TransmitUnencoded: pointer.Bool(true),
+					},
+				},
+			}
+
+			script, err := FilesToDiskScript(ctx, fakeClient, namespace, files)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(script).To(Equal(`
```

Comment:
We could adapt the test that it actually executes the script and writes the files into `os.MkdirTemp` directories to check their content.